### PR TITLE
Remove the txn field from TransactionAbortedError

### PR DESCRIPTION
**Pull request by kkaneda**
Introduce `Error.SetTxn` as we have to reset `Error.Message` when the `Txn`
field is set. Other than that, the change is similar to what we did to `TransactionRetryError`.

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/3952)

### CODE REVIEW

**Comment by vivekmenezes** [comment]
Review status: 0 of 12 files reviewed at latest revision, 1 unresolved discussion.

---

*[roachpb/errors.proto, line 74 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3952#-K8tF255BE5YATH8phe0:-K8tF255BE5YATH8phe1:41049984) ([raw file](https://github.com/cockroachdb/cockroach/blob/752c6b676778d096f21c54b1dcfc6db6ba5bce80/roachpb/errors.proto#L74)):*
This is empty. Do you plan on eliminating it in a separate PR?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3952)*

**Comment by kkaneda** [comment]
Review status: 0 of 12 files reviewed at latest revision, 1 unresolved discussion.

---

*[roachpb/errors.proto, line 74 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3952#-K8tF255BE5YATH8phe0:-K8tPEb_eS5zvGFKv8P5:831456250) ([raw file](https://github.com/cockroachdb/cockroach/blob/752c6b676778d096f21c54b1dcfc6db6ba5bce80/roachpb/errors.proto#L74)):*
Yes, I would like to, but that will require several iterations.. We still need this proto to restore the original Go type when converting a `roachpb.Error` back to a Go error.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3952)*

File: roachpb/errors.go [hunk 1]
```diff
@@ -172,6 +172,17 @@ func (e *Error) SetGoError(err error) {
 	}
 }
 
+// SetTxn sets the txn and resets the error message.
+func (e *Error) SetTxn(txn *Transaction) {
```
Comment (bdarnell) [inline roachpb/errors.go]: Can we give the field a custom name in the proto file to ensure that all access goes through setter methods (and a new getter)?

**Comment by bdarnell** [comment]
LGTM

**Comment by kkaneda** [comment]
Thanks for the review, @bdarnell !

File: roachpb/errors.go (hunk 1 above)
Comment (kkaneda) [inline roachpb/errors.go]: Ah, I see. First tried to un-expose the field, but got the following error from `go vet`

```
roachpb/errors.pb.go:354: struct field txn has json tag but is not exported
```

Will rename to something `UnexposedTxn`..

**Pull request merged**